{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [],
   "source": [
    "from random import choice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [],
   "source": [
    "bases = [\"A\", \"T\", \"C\", \"G\"]\n",
    "RNA_base = [\"A\", \"U\", \"C\", \"G\"]\n",
    "\n",
    "RNA_Codons = {\n",
    "    # 'M' - START, '_' - STOP\n",
    "    \"GCU\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \"GCG\": \"A\",\n",
    "    \"UGU\": \"C\", \"UGC\": \"C\",\n",
    "    \"GAU\": \"D\", \"GAC\": \"D\",\n",
    "    \"GAA\": \"E\", \"GAG\": \"E\",\n",
    "    \"UUU\": \"F\", \"UUC\": \"F\",\n",
    "    \"GGU\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\",\n",
    "    \"CAU\": \"H\", \"CAC\": \"H\",\n",
    "    \"AUA\": \"I\", \"AUU\": \"I\", \"AUC\": \"I\",\n",
    "    \"AAA\": \"K\", \"AAG\": \"K\",\n",
    "    \"UUA\": \"L\", \"UUG\": \"L\", \"CUU\": \"L\", \"CUC\": \"L\", \"CUA\": \"L\", \"CUG\": \"L\",\n",
    "    \"AUG\": \"M\",\n",
    "    \"AAU\": \"N\", \"AAC\": \"N\",\n",
    "    \"CCU\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\",\n",
    "    \"CAA\": \"Q\", \"CAG\": \"Q\",\n",
    "    \"CGU\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \"CGG\": \"R\", \"AGA\": \"R\", \"AGG\": \"R\",\n",
    "    \"UCU\": \"S\", \"UCC\": \"S\", \"UCA\": \"S\", \"UCG\": \"S\", \"AGU\": \"S\", \"AGC\": \"S\",\n",
    "    \"ACU\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n",
    "    \"GUU\": \"V\", \"GUC\": \"V\", \"GUA\": \"V\", \"GUG\": \"V\",\n",
    "    \"UGG\": \"W\",\n",
    "    \"UAU\": \"Y\", \"UAC\": \"Y\",\n",
    "    \"UAA\": \"_\", \"UAG\": \"_\", \"UGA\": \"_\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_raw_file(file):\n",
    "    \"\"\"\n",
    "    Reads a file containing the sequence(s)\n",
    "    \n",
    "    Parameter\n",
    "    ---------\n",
    "    file: string, name of the file, in the directory with text or fasta extention (.txt or .fasta)\n",
    "    \n",
    "    Returns\n",
    "    ---------\n",
    "    A list containing the content of the text file as string inside the list\n",
    "    \"\"\"\n",
    "    # Open the file \n",
    "    text = open(file, \"r\")\n",
    "    # Read it in a list\n",
    "    text_read = text.readlines()\n",
    "    text.close()\n",
    "    return text_read\n",
    "\n",
    "\n",
    "def check_nucleotide(seq_read):\n",
    "    \"\"\"\n",
    "    Takes the list file containing sequence(s) in string format\n",
    "    \n",
    "    Parameter\n",
    "    ---------\n",
    "    seq_read: list, provided by the read_rwa_file function\n",
    "    \n",
    "    Returns\n",
    "    ---------\n",
    "    A list containing clean sequence(s) in form of string with no white space and junck \n",
    "    elements in reads\n",
    "    \"\"\"\n",
    "    # Strip the potential white spaces around each read\n",
    "    # Fill the potential white space inside the sequence\n",
    "    # Turn each element to upper case\n",
    "    # Skip the reads staring with '>' for potential fasta format files\n",
    "    first = [x.strip().replace(\" \", \"\").upper() for x in seq_read if x.startswith(\">\") == False]\n",
    "    \n",
    "    # Check if there is any irrelevant element in each single read\n",
    "    # Elements should only be the ones in nuc list\n",
    "    second = [[x.replace(y, \"\") for x in y if x in bases] for y in first]\n",
    "    \n",
    "    # Join the processed read in one sinle piece \n",
    "    third = [\"\".join(x) for x in second]\n",
    "    return third[0]\n",
    "\n",
    "\n",
    "def transcribe(DNAseq):\n",
    "    \"\"\"\n",
    "    Transcribes the DNA sequence to RNA sequence \n",
    "    \n",
    "    Parameter\n",
    "    ---------\n",
    "    DNAseq: String, sequence provided by check_nucleotide function\n",
    "\n",
    "    Return\n",
    "    ---------\n",
    "    The transcribed RNA string\n",
    "    \"\"\"\n",
    "    return DNAseq.replace(\"T\", \"U\")\n",
    "\n",
    "\n",
    "def de_transcribe(RNAseq):\n",
    "    \"\"\"\n",
    "    De transcribes the RNA sequence to DNA sequence \n",
    "    \n",
    "    Parameter\n",
    "    ---------\n",
    "    RNAseq: String, sequence provided by transcribe function\n",
    "\n",
    "    Return\n",
    "    ---------\n",
    "    The de-transcribed DNA string\n",
    "    \"\"\"\n",
    "    return RNAseq.replace(\"U\", \"T\")\n",
    "\n",
    "\n",
    "def gene_detector(seq):\n",
    "    \"\"\"\n",
    "    Detects the gene in the sequence including starting and terminating codons\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    seq: String, sequence provided by check_nucleotide function\n",
    "    \n",
    "    Returns\n",
    "    ----------\n",
    "    A tuple containing the index of starting codon (int), the index of terminating codon (int) and\n",
    "    the gene (str)\n",
    "    \"\"\"\n",
    "    rna_seq = transcribe(seq)\n",
    "    # Starting codon\n",
    "    gene_start = \"AUG\"\n",
    "    # Terminating codons\n",
    "    gene_end = [\"UAA\", \"UAG\", \"UGA\"]\n",
    "    \n",
    "    # Finding which terminating codon is found first\n",
    "    ending = sorted([(x, rna_seq.index(x)) for x in gene_end if x in rna_seq])[0][0]\n",
    "    # Isolating gene from starting codon to the terminating codon from the whole_seq\n",
    "    start_codon_index = rna_seq.index(gene_start.upper())\n",
    "    term_codon_index = rna_seq.index(ending) + 3\n",
    "    rna = rna_seq[start_codon_index: term_codon_index]\n",
    "    \n",
    "    return start_codon_index, term_codon_index, rna\n",
    "\n",
    "\n",
    "def label_sites(seq, res_sites):\n",
    "    \"\"\"\n",
    "    Analyzes the restriction site(s) against the RNA sequence and adds a label to the \n",
    "    restriction site(s) depending on which part of RNA sequences it binds to.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    seq: String, sequence provided by check_nucleotide function\n",
    "    res_sites: List, containing all the restriction sites provided by the user\n",
    "    \n",
    "    Returns\n",
    "    ----------\n",
    "    A tuple whose first element is a list with restrction sites with labels and second\n",
    "    element is a dictionary of restriction sites that were not found in the sequence.\n",
    "    \"\"\"\n",
    "    position = []\n",
    "    # If restriction site is not in the sequence, it will be added to this dictionary\n",
    "    report = {}\n",
    "    # Translate DNA to RNA\n",
    "    rna_seq = transcribe(seq)\n",
    "    # Translate DNA restriction site(s) to RNA\n",
    "    reses = list(map(transcribe, res_sites))\n",
    "    \n",
    "    # Making a list of RNA restriction site - length of the site... \n",
    "    # ... - starting point in the main sequence - ending point in the main sequence\n",
    "    for res in reses:\n",
    "        try:\n",
    "            position.append([res, len(res), rna_seq.index(res), rna_seq.index(res) + len(res) - 1])\n",
    "        # If the restriction site is not found, it will be added to the report dictionary\n",
    "        except ValueError:\n",
    "            report[res] = \"This restriction site was not found.\"\n",
    "            \n",
    "    # Locating the position of the restriction site     \n",
    "    for pos in position:\n",
    "        # OUT-5: upstream of the starting codon (upstream of 5')\n",
    "        if (pos[2] and pos[3]) < start:\n",
    "            pos.append(\"Out-5\")\n",
    "        # OUT-3: downstream of the terminating codon (downstream of 3')\n",
    "        elif (pos[2] and pos[3]) > end - 1:\n",
    "            pos.append(\"Out-3\")\n",
    "        # Over-5: overlapping with the starting codon\n",
    "        elif (pos[2] < start) and (start < pos[3] < end - 1):\n",
    "            pos.append(\"Over-5\")\n",
    "        # Over-3: overlapping with the terminating codon\n",
    "        elif (pos[2] < end - 1) and (pos[3] > end - 1):\n",
    "            pos.append(\"Over-3\")\n",
    "        # Exon: restriction site is in the middle of the gene\n",
    "        else:\n",
    "            pos.append(\"Exon\")\n",
    "            \n",
    "    return (position, report)\n",
    "\n",
    "\n",
    "def exon_shuffler(seq, labeled_site):\n",
    "    \"\"\"\n",
    "    Finds where the restriction site lies in the exon segment of the gene. Then change the\n",
    "    codon with one its alternatives to change the sequence without changing the protein.\n",
    "    \n",
    "    Parameters\n",
    "    ---------\n",
    "    seq: String, sequence provided by check_nucleotide function\n",
    "    labeled_site: String, restriction site sequence that lies across the exon segment of the gene.\n",
    "    \n",
    "    Returns\n",
    "    ---------\n",
    "    Modified exon part of the gene with alternative codons\n",
    "    \"\"\"\n",
    "    \n",
    "    sites = labeled_site\n",
    "    start_codon_index, term_codon_index, gene = gene_detector(seq)\n",
    "    # Find the index where restriction site starts\n",
    "    rs_e = gene.find(sites[0])\n",
    "    rs_l = len(sites[0]) - 1\n",
    "    \n",
    "    # Turning RNA sequence into codon triplets\n",
    "    rna_triplet = [gene[i: i + 3] for i in range(0, len(gene), 3)]\n",
    "    # Creating a list of the index numbers where the restriction site lies\n",
    "    rest_site = [x for x in range(rs_e, rs_e + rs_l + 1)]\n",
    "\n",
    "    pairs = []\n",
    "\n",
    "    # Find with which codon triplets the restriction site aligns\n",
    "    for nuc in rest_site:\n",
    "        y = nuc / 3\n",
    "        if y.is_integer() == True:\n",
    "            pairs.append(int(y - 1))\n",
    "\n",
    "        else:\n",
    "            pairs.append(int(y))\n",
    "\n",
    "    pairs = list(set(pairs))        \n",
    "    \n",
    "    for pair in pairs:\n",
    "        triplet = rna_triplet[pair]\n",
    "        value = RNA_Codons[triplet]\n",
    "        # Detect the alternative codons for the codon triplet in the RNA sequence\n",
    "        other_codons = [k for k, v in RNA_Codons.items() if v == value and k != triplet]\n",
    "        # If there are more than one alternative, choose a triplet with random.choice\n",
    "        final_codon = choice(other_codons)\n",
    "        # Add the new alternative to the RNA codon triplet list\n",
    "        rna_triplet.insert(pair, final_codon)\n",
    "        # Remove the old codon triplet from the list\n",
    "        rna_triplet.pop(pair + 1)\n",
    "        \n",
    "    # Join the triplets to reconstruct the RNA sequence\n",
    "    final = \"\".join(rna_triplet)\n",
    "    return final    \n",
    "    \n",
    "    \n",
    "def sequence_modification(seq, res_sites):\n",
    "    \"\"\"\n",
    "    Modifies the sequence according to the existing restriction site.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    seq: String, DNA sequence provided by check_nucleotide function\n",
    "    res_sites: List, containing all the restriction sites provided by the user\n",
    "    \n",
    "    Returns:\n",
    "    ----------\n",
    "    String, the modified sequence\n",
    "    \"\"\"   \n",
    "    rna_seq = transcribe(seq)\n",
    "    sites = label_sites(seq, res_sites)\n",
    "    start_codon_index, term_codon_index, gene = gene_detector(seq)\n",
    "    \n",
    "    for pos in sites:\n",
    "        # If the tag is \"Out-5\", just remove any sequence upstream of that site\n",
    "        if pos[-1] == \"Out-5\":\n",
    "            rna_seq = rna_seq[pos[-3]:]\n",
    "        # If the tag is \"Out-3\", just remove any sequence downstream that site\n",
    "        elif pos[-1] == \"Out-3\":\n",
    "            rna_seq = rna_seq[: pos[-2] + 1]\n",
    "        # If the tag is \"Over-5\", change the two nucleotides upstream of the starting codon\n",
    "        elif pos[-1] == \"Over-5\":\n",
    "            b1 = rna_seq[start_codon_index - 1]\n",
    "            b1_rep_list = [x for x in RNA_base if x != b1]\n",
    "            b1_replcaement = choice(b1_rep_list)\n",
    "            b2 = rna_seq[start_codon_index - 2]\n",
    "            b2_rep_list = [x for x in RNA_base if x != b2]\n",
    "            b2_replacement = choice(b2_rep_list)\n",
    "            rna_seq = rna_seq[:start_codon_index - 2] + b2_replacement + b1_replcaement + rna_seq[start_codon_index:]\n",
    "        # If the tag is \"Over-3\", change the two nucleotides downstream of the terminating codon\n",
    "        elif pos[-1] == \"Over-3\":\n",
    "            b1 = rna_seq[term_codon_index]\n",
    "            b1_rep_list = [x for x in RNA_base if x != b1]\n",
    "            b1_replcaement = choice(b1_rep_list)\n",
    "            b2 = rna_seq[term_codon_index + 1]\n",
    "            b2_rep_list = [x for x in RNA_base if x != b2]\n",
    "            b2_replacement = choice(b2_rep_list)\n",
    "            rna_seq = rna_seq[:term_codon_index] + b1_replcaement + b2_replacement + rna_seq[term_codon_index + 2:]\n",
    "\n",
    "        elif pos[-1] == \"Exon\":\n",
    "            mod_gene = exon_shuffler(seq, pos)\n",
    "            rna_seq = rna_seq[:start_codon_index] + mod_gene + rna_seq[term_codon_index + 3:]\n",
    "            \n",
    "\n",
    "        else:\n",
    "            pass\n",
    "        \n",
    "        modified = de_transcribe(rna_seq)\n",
    "        \n",
    "    return modified"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
